File: lab6_1.kt
abstract class Country {
    abstract fun printName()
    abstract fun printCapital()
    abstract fun printFavFood()

    fun printCountryDetails() {
        printName()
        printCapital()
        printFavFood()
    }
}

class Romania : Country() {
    override fun printName() {
        println("Romania")
    }

    override fun printCapital() {
        println("Bucharest")
    }

    override fun printFavFood() {
        println("Sarmale \n")
    }
}

class France : Country() {
    override fun printName() {
        println("France")
    }

    override fun printCapital() {
        println("Paris")
    }

    override fun printFavFood() {
        println("Escargot \n")
    }
}

class Germany : Country() {
    override fun printName() {
        println("Deutschland")
    }

    override fun printCapital() {
        println("Berlin")
    }

    override fun printFavFood() {
        println("Sauerkraut \n")
    }
}

fun main(){
    var ro = Romania()
    ro.printCountryDetails()

    var fr = France()
    fr.printCountryDetails()

    var de = Germany()
    de.printCountryDetails()

}

File: labs/lab6/lab6_1.cpp
#include <iostream>
#include <string>
#include <cstdlib>

using namespace std;

class Country
{
public:
    virtual void print_name() = 0;
    virtual void print_capital() = 0;
    virtual void print_favourite_food() = 0;

    void print_country_details()
    {
        print_name();
        print_capital();
        print_favourite_food();
        std::cout << "\n";
    }
};

class Romania : public Country
{
public:
    void print_name()
    {
        std::cout << "Romania\n";
    }

    void print_capital()
    {
        std::cout << "Bucharest\n";
    }

    void print_favourite_food()
    {
        std::cout << "Sarmale\n";
    }
};

class France : public Country
{
    public:
    void print_name()
    {
        std::cout << "France\n";
    }

    void print_capital()
    {
        std::cout << "Paris\n";
    }

    void print_favourite_food()
    {
        std::cout << "Escargot\n";
    }
};

class Germany : public Country
{
    public:
    void print_name()
    {
        std::cout << "Deutschland\n";
    }

    void print_capital()
    {
        std::cout << "Berlin\n";
    }

    void print_favourite_food()
    {
        std::cout << "Sauerkraut\n";
    }
};

int main()
{
    Romania ro;
    France fr;
    Germany de;

    ro.print_country_details();
    fr.print_country_details();
    de.print_country_details();

    return 0;
}